tests/unit/pip_sync_faster/main_test.py

```diff
@@ -1,6 +1,156 @@
-from pip_sync_faster.main import hello_world
+import json
+from importlib.metadata import version
+from subprocess import CalledProcessError
 
+import pytest
 
-class TestHelloWorld:
-    def test_it(self):
-        assert hello_world() == "Hello, world!"
+from pip_sync_faster import main
+
+
+class TestPipSyncMaybe:
+    def test_if_the_hashes_match_it_doesnt_call_pip_sync(
+        self, cache_hashes, run, requirements_files
+    ):
+        # Make sure that all the cached hashes are present and matching.
+        cache_hashes(main.get_hashes(requirements_files))
+
+        main.pip_sync_maybe(requirements_files)
+
+        run.assert_not_called()
+
+    def test_if_theres_no_cached_hashes(
+        self, assert_hashes_cached, requirements_files, run
+    ):
+        main.pip_sync_maybe(requirements_files)
+
+        run.assert_called_once_with(["pip-sync", *requirements_files], check=True)
+        assert_hashes_cached(requirements_files)
+
+    def test_if_theres_one_wrong_hash(
+        self, assert_hashes_cached, cache_hashes, requirements_files, run
+    ):
+        # Make the cached hashes file contain one non-matching hash.
+        hashes = main.get_hashes(requirements_files)
+        hashes[list(hashes.keys())[0]] = "non_matching_hash"
+        cache_hashes(hashes)
+
+        main.pip_sync_maybe(requirements_files)
+
+        run.assert_called_once_with(["pip-sync", *requirements_files], check=True)
+        assert_hashes_cached(requirements_files)
+
+    def test_if_theres_one_missing_hash(
+        self, assert_hashes_cached, cache_hashes, requirements_files, run
+    ):
+        # Make the cached hashes file be missing one hash.
+        cache_hashes(main.get_hashes(requirements_files[1:]))
+
+        main.pip_sync_maybe(requirements_files)
+
+        run.assert_called_once_with(["pip-sync", *requirements_files], check=True)
+        assert_hashes_cached(requirements_files)
+
+    def test_if_theres_a_different_files_hash_cached(
+        self, assert_hashes_cached, cache_hashes, requirements_files, run
+    ):
+        # The cache contains correct hashes for both dev.txt and format.txt.
+        cache_hashes(main.get_hashes(requirements_files))
+
+        # pip-sync-faster is called with just format.txt.
+        main.pip_sync_maybe(requirements_files[1:])
+
+        # It should call pip-sync: the last time that pip-sync-faster was
+        # called (and updated the cache) it was called with both dev.txt and
+        # format.txt. Now it's being called with just format.txt. Any
+        # requirements that're in dev.txt but not in format.txt need to be
+        # removed from the venv.
+        run.assert_called_once_with(["pip-sync", *requirements_files[1:]], check=True)
+        # It should update the cache to contain only format.txt.
+        assert_hashes_cached(requirements_files[1:])
```

Comment:
Are the three tests cases above really different? Might be an opportunity for less here.